src/ir/module-utils.h

```diff
@@ -444,10 +461,57 @@ collectSignatures(Module& wasm,
       counts[innerPair.first] += innerPair.second;
     }
   }
+  // We must sort all the dependencies of a signature before it. For example,
+  // (func (param (ref (func)))) must appear after (func). To do that, find the
+  // depth of dependencies of each signature. For example, if A depends on B
+  // which depends on C, then A's depth is 2, B's is 1, and C's is 0 (assuming
+  // no other dependencies).
+  Counts depthOfDependencies;
+  std::unordered_map<Signature, std::unordered_set<Signature>> isDependencyOf;
+  // To calculate the depth of dependencies, we'll do a flow analysis, visiting
+  // each signature as we find out new things about it.
+  std::set<Signature> toVisit;
+  for (auto& pair : counts) {
+    auto sig = pair.first;
+    depthOfDependencies[sig] = 0;
+    toVisit.insert(sig);
+    for (Type type : {sig.params, sig.results}) {
+      for (auto subType : type) {
+        if (subType.isRef()) {
+          auto heapType = subType.getHeapType();
+          if (heapType.isSignature()) {
+            isDependencyOf[heapType.getSignature()].insert(sig);
+          }
+        }
+      }
+    }
+  }
+  while (!toVisit.empty()) {
+    auto iter = toVisit.begin();
+    auto sig = *iter;
+    toVisit.erase(iter);
+    // Anything that depends on this has a depth of dependencies equal to this
+    // signature's, plus this signature itself.
+    auto newDepth = depthOfDependencies[sig] + 1;
+    if (newDepth > counts.size()) {
+      Fatal() << "Cyclic signatures detected, cannot sort them.";
+    }
+    for (auto& other : isDependencyOf[sig]) {
+      if (depthOfDependencies[other] < newDepth) {
+        // We found something new to propagate.
+        depthOfDependencies[other] = newDepth;
+        toVisit.insert(other);
+      }
+    }
+  }
+  // Sort by frequency and then simplicity, and also keeping every signature
+  // before things that depend on it.
   std::vector<std::pair<Signature, size_t>> sorted(counts.begin(),
                                                    counts.end());
   std::sort(sorted.begin(), sorted.end(), [&](auto a, auto b) {
-    // order by frequency then simplicity
+    if (depthOfDependencies[a.first] != depthOfDependencies[b.first]) {
+      return depthOfDependencies[a.first] < depthOfDependencies[b.first];
+    }
```

Comment:
It would be more optimal to calculate the transitive closure of the dependency graph and use that directly in the sort function rather than approximating it with dependency path lengths, but perhaps it doesn't matter too much if we're going to remove all this code once we support building recursive types.

Are there other places where we need to calculate dependency graphs or topological sorts? I'm wondering if it would be worthwhile to pull this out as a generic graph processing routine.

Reply:
Talked about this offline - the missing piece was that it is important for this code to order types by usage to minimize the cumulative size of lebs in type annotations. Once we support recursive types (and more generally, once we support referring to undefined types in type definitions) this ordering code will no longer be necessary.